crypto/fipsmodule/sha/sha512.c

```diff
@@ -288,13 +333,21 @@ static int sha512_final_impl(uint8_t *out, size_t md_len, SHA512_CTX *sha) {
     return 0;
   }
 
-  assert(md_len % 8 == 0);
+  // TODO [childw] explain this
   const size_t out_words = md_len / 8;
+  const size_t remainder = md_len % 8;
+  assert(remainder == 0 || (remainder == 4 && out_words * 8 + remainder == md_len));
   for (size_t i = 0; i < out_words; i++) {
     CRYPTO_store_u64_be(out, sha->h[i]);
     out += 8;
   }
 
+  // TODO [childw] explain this
+  if (remainder == 4) {
+    const uint64_t trailer = CRYPTO_bswap8(sha->h[out_words]);
```

Comment:
Do you always need to swap here, or should it use `CRYPTO_store_u64_be` which swaps on little endian platform and does not swap on big endian platforms (pending https://github.com/aws/aws-lc/pull/1133).

Reply:
Hm, good question. In the current small-endian world, we want to take the rightmost 4 bytes of the last element of the state array, swap the order of those bytes, and store them in the last 4 bytes of the output buffer. That byte order swap would not be required on big-endian systems. However, we can't use `CRYPTO_store_u64_be` here to write directly to the output buffer, as that would potentially overwrite the end of the output buffer by 4 bytes (SHA512-224 digest size is 28 bytes, which is not 8-byte-word-aligned).

To avoid `#ifdef`'ing here, I'll use `CRYPTO_store_u64_be` to store the full 8-byte word into a temporary variable and then only write the first 4 bytes of that to the output buffer.